elementary/monitor/alerts/group_of_alerts.py

```diff
@@ -0,0 +1,337 @@
+from dataclasses import dataclass
+from datetime import datetime
+from enum import Enum
+from typing import Dict, List
+
+from elementary.monitor.alerts.alert import Alert, SlackAlertMessageBuilder
+from elementary.monitor.alerts.model import ModelAlert
+from elementary.monitor.fetchers.alerts.normalized_alert import CHANNEL_KEY
+from elementary.utils.json_utils import try_load_json
+from elementary.utils.models import alert_to_concise_name, get_shortened_model_name
+
+DATETIME_FORMAT = "%Y-%m-%d %H:%M:%S"
+
+
+class GroupingType(Enum):
+    BY_ALERT = "alert"
+    BY_TABLE = "table"
+    ALL = "all"
+
+
+@dataclass(
+    frozen=True, eq=True
+)  # frozen+eq defined so we can use it as a dict key. Also, it's all Strings
+class NotificationComponent:
+    name_in_summary: str
+    empty_section_content: str
+
+
+@dataclass(frozen=True, eq=True)
+class AlertGroupComponent(NotificationComponent):
+    emoji_in_summary: str
+    name_in_full: str
+    emoji_in_full: str
+
+
+ErrorComponent = AlertGroupComponent(
+    name_in_summary="Errors",
+    emoji_in_summary="exclamation",
+    name_in_full="Error",
+    emoji_in_full="exclamation",
+    empty_section_content="No Errors",
+)
+
+WarningComponent = AlertGroupComponent(
+    name_in_summary="Warning",
+    emoji_in_summary="warning",
+    name_in_full="Warning",
+    emoji_in_full="warning",
+    empty_section_content="No Warnings",
+)
+
+FailureComponent = AlertGroupComponent(
+    name_in_summary="Failed",
+    emoji_in_summary="small_red_triangle",
+    name_in_full="Failed tests",
+    emoji_in_full="X",
+    empty_section_content="No Failures",
+)
+
+
+TagsComponent = NotificationComponent(
+    name_in_summary="Tags", empty_section_content="No Tags"
+)
+OwnersComponent = NotificationComponent(
+    name_in_summary="Owners", empty_section_content="No Owners"
+)
+SubsComponent = NotificationComponent(
+    name_in_summary="Subscribers", empty_section_content="No Subscribers"
+)
+
+
+class GroupOfAlerts(SlackAlertMessageBuilder):
+    def __init__(self, alerts: List[Alert], default_channel_destination: str):
+
+        self.alerts = alerts
+
+        self._sort_channel_destination(default_channel=default_channel_destination)
+        self._fill_components_to_alerts(alerts)
+
+        tags = self._fill_and_dedup_tags(alerts)
+        self._components_to_attn_required: Dict[NotificationComponent, str] = {
+            TagsComponent: tags
+        }
+        # self_components_to_attn_required is a magic dict that maps:
```

Comment:
I got self.owners and self.subscribers that needs to be updated from outside this class (because of the Slack API usage)
I also need them mapped from their components, for the sake of the dict. 
I can have avoid the usage of magic dicts but that would make the code have a hard-set relationship between the component call OwnersComponent and the attr called self.owners. I used the magic dict to avoid it, but really, we can just have it.

Reply:
decided together to just replace self.owners with self._owners and expose ```   def get_owners(self):
        ...
    def set_onwers(self):
        ... ```